src/machine/aot/aot.x64.c

```diff
@@ -0,0 +1,1284 @@
+
+  switch (b.tag) {
+    case AOT_TAG_REGISTER:
+      | op2_x_r imul, rax, b.value.reg
+      break;
+    case AOT_TAG_IMMEDIATE:
+      | load_imm rcx, b.value.i
+      | imul rax, rcx
+      break;
+    case AOT_TAG_X64_REGISTER:
+      | imul rax, Rq(b.value.x64_reg)
+      break;
+  }
+
+  | op2_r_x mov, target, rax
+
+  return DASM_S_OK;
+}
+
+int aot_mulh(AotContext* context, riscv_register_t target, AotValue a, AotValue b, int is_signed)
+{
+  int ret;
+  uint32_t loc1;
+  dasm_State** Dst = &context->d;
+
+  ret = aot_mov_x64(context, X64_RAX, a);
+  if (ret != DASM_S_OK) { return ret; }
+
+  switch (b.tag) {
+    case AOT_TAG_REGISTER:
+      if (is_signed) {
+        | op1_r imul, b.value.reg
+      } else {
+        | op1_r mul, b.value.reg
+      }
+      break;
+    case AOT_TAG_IMMEDIATE:
+      | load_imm rcx, b.value.i
+      if (is_signed) {
+        | imul rcx
+      } else {
+        | mul rcx
+      }
+      break;
+    case AOT_TAG_X64_REGISTER:
+      if (is_signed) {
+        | imul Rq(b.value.x64_reg)
+      } else {
+        | mul Rq(b.value.x64_reg)
+      }
+      break;
+  }
+  | op2_r_x mov, target, rdx
+
+  return DASM_S_OK;
+}
+
+/* Inspired from https://github.com/rv8-io/rv8/blob/834259098a5c182874aac97d82a164d144244e1a/src/jit/jit-emitter-rv64.h#L931 */
+int aot_mulhsu(AotContext* context, riscv_register_t target, AotValue a, AotValue b)
+{
+  int ret;
+  uint32_t loc1;
+  dasm_State** Dst = &context->d;
+
+  ret = aot_mov_x64(context, X64_RAX, a);
+  if (ret != DASM_S_OK) { return ret; }
+
+  | test rax, rax
+  | jns >1
+  /* calculate res = mulhu(-a, b), res is stored in rdx after this. */
+  | neg rax
+  switch (b.tag) {
+    case AOT_TAG_REGISTER:
+      | op1_r mul, b.value.reg
+      break;
+    case AOT_TAG_IMMEDIATE:
+      | load_imm rcx, b.value.i
+      | mul rcx
+      break;
+    case AOT_TAG_X64_REGISTER:
+      | mul Rq(b.value.x64_reg)
+      break;
+  }
+  /* calculate ~res and store it in rcx */
+  | xor rdx, -1
+  | mov rcx, rdx
+  /*
+   * calculate (a * b), then test (a * b == 0) and convert that to 1 or 0,
+   * result is stored in rax after this.
+   */
+  ret = aot_mov_x64(context, X64_RAX, a);
+  if (ret != DASM_S_OK) { return ret; }
+  switch (b.tag) {
+    case AOT_TAG_REGISTER:
+      | op2_x_r imul, rax, b.value.reg
+      break;
+    case AOT_TAG_IMMEDIATE:
+      | load_imm rdx, b.value.i
+      | imul rax, rdx
+      break;
+    case AOT_TAG_X64_REGISTER:
+      | imul rax, Rq(b.value.x64_reg)
+      break;
+  }
+  | test rax, rax
+  | setz al
+  | movzx rax, al
+  /* calculate ~res + (a * b == 0) */
+  | add rax, rcx
+  | jmp >2
+  /* just mulhu here */
+  |1:
+  switch (b.tag) {
+    case AOT_TAG_REGISTER:
+      | op1_r mul, b.value.reg
+      break;
+    case AOT_TAG_IMMEDIATE:
+      | load_imm rcx, b.value.i
+      | mul rcx
+      break;
+    case AOT_TAG_X64_REGISTER:
+      | mul Rq(b.value.x64_reg)
+      break;
+  }
+  /* calculate ~res and store it in rcx */
```

Comment:
There is no rcx in below code.
Is this comment correct?

Reply:
Oops this is something I forgot to delete